File: solutions/javascript/pig-latin/1/pig-latin.js
export const translate = (phrase) =>
  phrase
    .split(" ")
    .map((word) => translateWord(word))
    .join(" ");

const translateWord = (original) => {
  // Starts with vowel or 'xr' or 'yt'
  let match = original.match(/^([aeiou]|xr|yt)/);
  if (match) return original + "ay";

  // Handle 'qu'
  match = original.match(/^(.*qu)(.*)/);
  if (match) return match[2] + match[1] + "ay";

  // Handle 'y'
  match = original.match(/^([^aeiou]+)(y.*)/);
  if (match) return match[2] + match[1] + "ay";

  // Handle rest
  match = original.match(/(^[^aeiou]+)(.*)/);
  if (match) return match[2] + match[1] + "ay";
};

File: solutions/python/pig-latin/1/pig_latin.py
import re


def translate(text):
    return " ".join([translate_word(word) for word in text.split()])


def translate_word(word):
    starts_with_vowel = re.compile(r"^([aeiou]|xr|yt)")
    if starts_with_vowel.search(word):
        print(starts_with_vowel.match(word))
        return glue_parts(word)

    has_qu = re.compile(r"^(?P<start>.*qu)(?P<end>.*)")
    has_y_within = re.compile(r"^(?P<start>[^aeiou]+)(?P<end>y.*)")
    starts_with_consonants = re.compile(r"^(?P<start>[^aeiou]+)(?P<end>.*)")

    for check in [has_qu, has_y_within, starts_with_consonants]:
        parts = check.match(word)
        if parts:
            return glue_parts(parts.group('end'), parts.group('start'))


def glue_parts(part1, part2=''):
    return part1 + part2 + 'ay'
